File: ios/Classes/SmileIDPlugin.swift
import Flutter
import SmileID
import UIKit

public class SmileIDPlugin: NSObject, FlutterPlugin, SmileIDApi {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger: FlutterBinaryMessenger = registrar.messenger()
        let api: SmileIDApi & NSObjectProtocol = SmileIDPlugin()
        SmileIDApiSetup.setUp(binaryMessenger: messenger, api: api)
        
        let documentVerificationFactory = SmileIDDocumentVerification.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            documentVerificationFactory,
            withId: SmileIDDocumentVerification.VIEW_TYPE_ID
        )
        
        let enhancedDocumentVerificationFactory = SmileIDEnhancedDocumentVerification.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            enhancedDocumentVerificationFactory,
            withId: SmileIDEnhancedDocumentVerification.VIEW_TYPE_ID
        )
        
        let smartSelfieEnrollmentFactory = SmileIDSmartSelfieEnrollment.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            smartSelfieEnrollmentFactory,
            withId: SmileIDSmartSelfieEnrollment.VIEW_TYPE_ID
        )
        
        let smartSelfieAuthenticationFactory = SmileIDSmartSelfieAuthentication.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            smartSelfieAuthenticationFactory,
            withId: SmileIDSmartSelfieAuthentication.VIEW_TYPE_ID
        )

         let smartSelfieEnrollmentEnhancedFactory = SmileIDSmartSelfieEnrollmentEnhanced.Factory(
             messenger: registrar.messenger()
         )
         registrar.register(
             smartSelfieEnrollmentEnhancedFactory,
             withId: SmileIDSmartSelfieEnrollmentEnhanced.VIEW_TYPE_ID
         )

         let smartSelfieAuthenticationEnhancedFactory = SmileIDSmartSelfieAuthenticationEnhanced.Factory(
             messenger: registrar.messenger()
         )
         registrar.register(
             smartSelfieAuthenticationEnhancedFactory,
             withId: SmileIDSmartSelfieAuthenticationEnhanced.VIEW_TYPE_ID
         )
        
        let biometricKYCFactory = SmileIDBiometricKYC.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            biometricKYCFactory,
            withId: SmileIDBiometricKYC.VIEW_TYPE_ID
        )
        
        let selfieCaptureFactory = SmileIDSmartSelfieCaptureView.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            selfieCaptureFactory,
            withId: SmileIDSmartSelfieCaptureView.VIEW_TYPE_ID
        )
        
        let documentCaptureFactory = SmileIDDocumentCaptureView.Factory(
            messenger: registrar.messenger()
        )
        registrar.register(
            documentCaptureFactory,
            withId: SmileIDDocumentCaptureView.VIEW_TYPE_ID
        )
    }
    
    func initializeWithApiKey(
        apiKey: String,
        config: FlutterConfig,
        useSandbox: Bool,
        enableCrashReporting: Bool
    ) {
        SmileID.initialize(
            apiKey: apiKey,
            config: config.toRequest(),
            useSandbox: useSandbox
        )
    }
    
    func initializeWithConfig(
        config: FlutterConfig,
        useSandbox: Bool,
        enableCrashReporting: Bool
    ) {
        SmileID.initialize(
            config: config.toRequest(),
            useSandbox: useSandbox
        )
    }
    
    func initialize(
        useSandbox: Bool
    ) {
        SmileID.initialize(
            useSandbox: useSandbox
        )
    }
    
    func setCallbackUrl(callbackUrl: String) {
        SmileID.setCallbackUrl(url: URL(string: callbackUrl))
    }
    
    func setAllowOfflineMode(allowOfflineMode: Bool) {
        SmileID.setAllowOfflineMode(allowOfflineMode: allowOfflineMode)
    }
    
    func getSubmittedJobs() -> [String] {
        SmileID.getSubmittedJobs()
    }
    
    func getUnsubmittedJobs() -> [String] {
        SmileID.getUnsubmittedJobs()
    }
    
    func cleanup(jobId: String) throws {
        try SmileID.cleanup(jobId: jobId)
    }
    
    func cleanupJobs(jobIds: [String]) throws {
        try SmileID.cleanup(jobIds: jobIds)
    }
    
    func submitJob(jobId: String, deleteFilesOnSuccess: Bool) throws {
        try SmileID.submitJob(jobId: jobId, deleteFilesOnSuccess: deleteFilesOnSuccess)
    }
    
    func authenticate(
        request: FlutterAuthenticationRequest,
        completion: @escaping (Result<FlutterAuthenticationResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.authenticate(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func prepUpload(
        request: FlutterPrepUploadRequest,
        completion: @escaping (Result<FlutterPrepUploadResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.prepUpload(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func upload(
        url : String,
        request : FlutterUploadRequest,
        completion: @escaping (Result<Void, Error>) -> Void
    ) {
        Task {
            do {
                let zipData = try request.toRequest()
                try await SmileID.api.upload(zip: zipData,to : url)
                completion(.success(()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func doEnhancedKyc(
        request: FlutterEnhancedKycRequest,
        completion: @escaping (Result<FlutterEnhancedKycResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.doEnhancedKyc(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func doEnhancedKycAsync(
        request: FlutterEnhancedKycRequest,
        completion: @escaping (Result<FlutterEnhancedKycAsyncResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.doEnhancedKycAsync(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func doSmartSelfieEnrollment(
        signature: String,
        timestamp: String,
        selfieImage: String,
        livenessImages: [String],
        userId: String,
        partnerParams: [String?: String?]?,
        callbackUrl: String?,
        sandboxResult: Int64?,
        allowNewEnroll: Bool?,
        completion: @escaping (Result<FlutterSmartSelfieResponse, any Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.doSmartSelfieEnrollment(
                    signature: signature,
                    timestamp: timestamp,
                    selfieImage: MultipartBody(
                        withImage: getFile(atPath: selfieImage)!,
                        forKey: URL(fileURLWithPath: selfieImage).lastPathComponent,
                        forName: URL(fileURLWithPath: selfieImage).lastPathComponent
                    )!,
                    livenessImages: livenessImages.map {
                        MultipartBody(
                            withImage: getFile(atPath: $0)!,
                            forKey: URL(fileURLWithPath: $0).lastPathComponent,
                            forName: URL(fileURLWithPath: $0).lastPathComponent
                        )!
                    },
                    userId: userId,
                    partnerParams: convertNullableMapToNonNull(data: partnerParams),
                    callbackUrl: callbackUrl,
                    sandboxResult: sandboxResult.map { Int($0) },
                    allowNewEnroll: allowNewEnroll,
                    failureReason: nil,
                    metadata: Metadata.default()
                )
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func doSmartSelfieAuthentication(
        signature: String,
        timestamp: String,
        selfieImage: String,
        livenessImages: [String],
        userId: String,
        partnerParams: [String?: String?]?,
        callbackUrl: String?,
        sandboxResult: Int64?,
        completion: @escaping (Result<FlutterSmartSelfieResponse, any Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.doSmartSelfieAuthentication(
                    signature: signature,
                    timestamp: timestamp,
                    userId: userId,
                    selfieImage: MultipartBody(
                        withImage: getFile(atPath: selfieImage)!,
                        forKey: URL(fileURLWithPath: selfieImage).lastPathComponent,
                        forName: URL(fileURLWithPath: selfieImage).lastPathComponent
                    )!,
                    livenessImages: livenessImages.map {
                        MultipartBody(
                            withImage: getFile(atPath: $0)!,
                            forKey: URL(fileURLWithPath: $0).lastPathComponent,
                            forName: URL(fileURLWithPath: $0).lastPathComponent
                        )!
                    },
                    partnerParams: convertNullableMapToNonNull(data: partnerParams),
                    callbackUrl: callbackUrl,
                    sandboxResult: sandboxResult.map { Int($0) },
                    failureReason: nil,
                    metadata: Metadata.default()
                )
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getSmartSelfieJobStatus(
        request: FlutterJobStatusRequest,
        completion: @escaping (Result<FlutterSmartSelfieJobStatusResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response: SmartSelfieJobStatusResponse = try await SmileID.api.getJobStatus(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        completion: @escaping (Result<FlutterDocumentVerificationJobStatusResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response: DocumentVerificationJobStatusResponse = try await SmileID.api.getJobStatus(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getBiometricKycJobStatus(
        request: FlutterJobStatusRequest,
        completion: @escaping (Result<FlutterBiometricKycJobStatusResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response: BiometricKycJobStatusResponse = try await SmileID.api.getJobStatus(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getEnhancedDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        completion: @escaping (Result<FlutterEnhancedDocumentVerificationJobStatusResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response: EnhancedDocumentVerificationJobStatusResponse = try await SmileID.api.getJobStatus(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getProductsConfig(
        request: FlutterProductsConfigRequest,
        completion: @escaping (Result<FlutterProductsConfigResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.getProductsConfig(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getValidDocuments(
        request: FlutterProductsConfigRequest,
        completion: @escaping (Result<FlutterValidDocumentsResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.getValidDocuments(request: request.toRequest())
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func getServices(
        completion: @escaping (Result<FlutterServicesResponse, Error>) -> Void
    ) {
        Task {
            do {
                let response = try await SmileID.api.getServices()
                completion(.success(response.toResponse()))
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func pollSmartSelfieJobStatus(
        request: FlutterJobStatusRequest,
        interval: Int64,
        numAttempts: Int64,
        completion: @escaping (Result<FlutterSmartSelfieJobStatusResponse, Error>) -> Void
    ) {
        pollJobStatus(
            apiCall: SmileID.api.pollSmartSelfieJobStatus,
            request: request.toRequest(),
            interval: interval,
            numAttempts: numAttempts,
            completion: { result in
                switch result {
                case let .success(response):
                    completion(.success(response.toResponse()))
                case let .failure(error):
                    completion(.failure(error))
                }
            }
        )
    }
    
    func pollDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        interval: Int64,
        numAttempts: Int64,
        completion: @escaping (Result<FlutterDocumentVerificationJobStatusResponse, Error>) -> Void
    ) {
        pollJobStatus(
            apiCall: SmileID.api.pollDocumentVerificationJobStatus,
            request: request.toRequest(),
            interval: interval,
            numAttempts: numAttempts,
            completion: { result in
                switch result {
                case let .success(response):
                    completion(.success(response.toResponse()))
                case let .failure(error):
                    completion(.failure(error))
                }
            }
        )
    }
    
    func pollBiometricKycJobStatus(
        request: FlutterJobStatusRequest,
        interval: Int64,
        numAttempts: Int64,
        completion: @escaping (Result<FlutterBiometricKycJobStatusResponse, Error>) -> Void
    ) {
        pollJobStatus(
            apiCall: SmileID.api.pollBiometricKycJobStatus,
            request: request.toRequest(),
            interval: interval,
            numAttempts: numAttempts,
            completion: { result in
                switch result {
                case let .success(response):
                    completion(.success(response.toResponse()))
                case let .failure(error):
                    completion(.failure(error))
                }
            }
        )
    }
    
    func pollEnhancedDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        interval: Int64,
        numAttempts: Int64,
        completion: @escaping (Result<FlutterEnhancedDocumentVerificationJobStatusResponse, Error>) -> Void
    ) {
        pollJobStatus(
            apiCall: SmileID.api.pollEnhancedDocumentVerificationJobStatus,
            request: request.toRequest(),
            interval: interval,
            numAttempts: numAttempts,
            completion: { result in
                switch result {
                case let .success(response):
                    completion(.success(response.toResponse()))
                case let .failure(error):
                    completion(.failure(error))
                }
            }
        )
    }
    
    func pollJobStatus<RequestType, T: JobResult>(
        apiCall: @escaping (RequestType, TimeInterval, Int) async throws -> AsyncThrowingStream<JobStatusResponse<T>, Error>,
        request: RequestType,
        interval: Int64,
        numAttempts: Int64,
        completion: @escaping (Result<JobStatusResponse<T>, Error>) -> Void
    ) {
        Task {
            do {
                let timeInterval = convertToTimeInterval(milliSeconds: interval)
                guard let numAttemptsInt = Int(exactly: numAttempts) else {
                    completion(.failure(NSError(domain: "Invalid numAttempts value", code: -1, userInfo: nil)))
                    return
                }
                
                let pollStream = try await apiCall(request, timeInterval, numAttemptsInt)
                var result: JobStatusResponse<T>? = nil
                
                for try await res in pollStream {
                    result = res
                }
                
                if let finalResult = result {
                    completion(.success(finalResult))
                } else {
                    completion(.failure(NSError(domain: "Polling completed without a result", code: -1, userInfo: nil)))
                }
            } catch {
                completion(.failure(error))
            }
        }
    }
    
    func convertToTimeInterval(milliSeconds: Int64) -> TimeInterval {
        let seconds = milliSeconds / 1000
        return TimeInterval(seconds)
    }
}

File: android/src/main/kotlin/com/smileidentity/flutter/SmileIDPlugin.kt
package com.smileidentity.flutter

import FlutterAuthenticationRequest
import FlutterAuthenticationResponse
import FlutterBiometricKycJobStatusResponse
import FlutterConfig
import FlutterDocumentVerificationJobStatusResponse
import FlutterEnhancedDocumentVerificationJobStatusResponse
import FlutterEnhancedKycAsyncResponse
import FlutterEnhancedKycRequest
import FlutterEnhancedKycResponse
import FlutterJobStatusRequest
import FlutterPrepUploadRequest
import FlutterPrepUploadResponse
import FlutterProductsConfigRequest
import FlutterProductsConfigResponse
import FlutterServicesResponse
import FlutterSmartSelfieJobStatusResponse
import FlutterSmartSelfieResponse
import FlutterUploadRequest
import FlutterValidDocumentsResponse
import SmileIDApi
import android.app.Activity
import android.content.Context
import com.smileidentity.SmileID
import com.smileidentity.SmileIDOptIn
import com.smileidentity.flutter.enhanced.SmileIDSmartSelfieAuthenticationEnhanced
import com.smileidentity.flutter.enhanced.SmileIDSmartSelfieEnrollmentEnhanced
import com.smileidentity.metadata.models.WrapperSdkName
import com.smileidentity.networking.asFormDataPart
import com.smileidentity.networking.pollBiometricKycJobStatus
import com.smileidentity.networking.pollDocumentVerificationJobStatus
import com.smileidentity.networking.pollEnhancedDocumentVerificationJobStatus
import com.smileidentity.networking.pollSmartSelfieJobStatus
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import java.io.File
import java.net.URL
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.single
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class SmileIDPlugin :
    FlutterPlugin,
    SmileIDApi,
    ActivityAware {
    private var activity: Activity? = null
    private lateinit var appContext: Context
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO)

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        SmileIDApi.setUp(flutterPluginBinding.binaryMessenger, this)
        appContext = flutterPluginBinding.applicationContext

        // Set wrapper info for Flutter SDK
        try {
            val version = com.smileidentity.flutter.BuildConfig.SMILE_ID_VERSION
            SmileID.setWrapperInfo(WrapperSdkName.Flutter, version)
        } catch (e: Exception) {
            // Fallback to default version if BuildConfig is not available
            SmileID.setWrapperInfo(WrapperSdkName.Flutter, "11.0.0")
        }

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDDocumentVerification.VIEW_TYPE_ID,
            SmileIDDocumentVerification.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDSmartSelfieEnrollment.VIEW_TYPE_ID,
            SmileIDSmartSelfieEnrollment.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDSmartSelfieAuthentication.VIEW_TYPE_ID,
            SmileIDSmartSelfieAuthentication.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDSmartSelfieEnrollmentEnhanced.VIEW_TYPE_ID,
            SmileIDSmartSelfieEnrollmentEnhanced.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDSmartSelfieAuthenticationEnhanced.VIEW_TYPE_ID,
            SmileIDSmartSelfieAuthenticationEnhanced.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDBiometricKYC.VIEW_TYPE_ID,
            SmileIDBiometricKYC.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDEnhancedDocumentVerification.VIEW_TYPE_ID,
            SmileIDEnhancedDocumentVerification.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDSmartSelfieCaptureView.VIEW_TYPE_ID,
            SmileIDSmartSelfieCaptureView.Factory(flutterPluginBinding.binaryMessenger),
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            SmileIDDocumentCaptureView.VIEW_TYPE_ID,
            SmileIDDocumentCaptureView.Factory(flutterPluginBinding.binaryMessenger),
        )
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        SmileIDApi.setUp(binding.binaryMessenger, null)
    }

    override fun initializeWithApiKey(
        apiKey: String,
        config: FlutterConfig,
        useSandbox: Boolean,
        enableCrashReporting: Boolean,
    ) {
        SmileID.initialize(
            context = appContext,
            apiKey = apiKey,
            config = config.toRequest(),
            useSandbox = useSandbox,
            enableCrashReporting = false,
        )
    }

    override fun initializeWithConfig(
        config: FlutterConfig,
        useSandbox: Boolean,
        enableCrashReporting: Boolean,
    ) {
        SmileID.initialize(
            context = appContext,
            config = config.toRequest(),
            useSandbox = useSandbox,
            enableCrashReporting = false,
        )
    }

    override fun initialize(useSandbox: Boolean) {
        SmileID.initialize(
            context = appContext,
            useSandbox = useSandbox,
        )
    }

    override fun setCallbackUrl(callbackUrl: String) {
        SmileID.setCallbackUrl(callbackUrl = URL(callbackUrl))
    }

    override fun setAllowOfflineMode(allowOfflineMode: Boolean) {
        SmileID.setAllowOfflineMode(allowOfflineMode = allowOfflineMode)
    }

    override fun getSubmittedJobs(): List<String> = SmileID.getSubmittedJobs()

    override fun getUnsubmittedJobs(): List<String> = SmileID.getUnsubmittedJobs()

    override fun cleanup(jobId: String) {
        SmileID.cleanup(jobId = jobId)
    }

    override fun cleanupJobs(jobIds: List<String>) {
        SmileID.cleanup(jobIds = jobIds)
    }

    override fun submitJob(jobId: String, deleteFilesOnSuccess: Boolean) {
        scope.launch {
            SmileID.submitJob(jobId = jobId, deleteFilesOnSuccess = deleteFilesOnSuccess)
        }
    }

    override fun authenticate(
        request: FlutterAuthenticationRequest,
        callback: (Result<FlutterAuthenticationResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.authenticate(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun prepUpload(
        request: FlutterPrepUploadRequest,
        callback: (Result<FlutterPrepUploadResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.prepUpload(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun upload(
        url: String,
        request: FlutterUploadRequest,
        callback: (Result<Unit>) -> Unit,
    ) = launch(
        work = { SmileID.api.upload(url, request.toRequest()) },
        callback = callback,
    )

    override fun doEnhancedKyc(
        request: FlutterEnhancedKycRequest,
        callback: (Result<FlutterEnhancedKycResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.doEnhancedKyc(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun doEnhancedKycAsync(
        request: FlutterEnhancedKycRequest,
        callback: (Result<FlutterEnhancedKycAsyncResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.doEnhancedKycAsync(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun getSmartSelfieJobStatus(
        request: FlutterJobStatusRequest,
        callback: (Result<FlutterSmartSelfieJobStatusResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.getSmartSelfieJobStatus(request.toRequest()).toResponse() },
        callback = callback,
    )

    @OptIn(SmileIDOptIn::class)
    override fun doSmartSelfieEnrollment(
        signature: String,
        timestamp: String,
        selfieImage: String,
        livenessImages: List<String>,
        userId: String,
        partnerParams: Map<String?, String?>?,
        callbackUrl: String?,
        sandboxResult: Long?,
        allowNewEnroll: Boolean?,
        callback: (Result<FlutterSmartSelfieResponse>) -> Unit,
    ) = launch(
        work = {
            SmileID.api
                .doSmartSelfieEnrollment(
                    userId = userId,
                    selfieImage =
                    File(selfieImage).asFormDataPart(
                        partName = "selfie_image",
                        mediaType = "image/jpeg",
                    ),
                    livenessImages =
                    livenessImages.map {
                        File(selfieImage).asFormDataPart(
                            partName = "liveness_images",
                            mediaType = "image/jpeg",
                        )
                    },
                    partnerParams = convertNullableMapToNonNull(partnerParams),
                    callbackUrl = callbackUrl,
                    sandboxResult = sandboxResult?.toInt(),
                    allowNewEnroll = allowNewEnroll,
                ).toResponse()
        },
        callback = callback,
    )

    @OptIn(SmileIDOptIn::class)
    override fun doSmartSelfieAuthentication(
        signature: String,
        timestamp: String,
        selfieImage: String,
        livenessImages: List<String>,
        userId: String,
        partnerParams: Map<String?, String?>?,
        callbackUrl: String?,
        sandboxResult: Long?,
        callback: (Result<FlutterSmartSelfieResponse>) -> Unit,
    ) = launch(
        work = {
            SmileID.api
                .doSmartSelfieAuthentication(
                    userId = userId,
                    selfieImage =
                    File(selfieImage).asFormDataPart(
                        partName = "selfie_image",
                        mediaType = "image/jpeg",
                    ),
                    livenessImages =
                    livenessImages.map {
                        File(selfieImage).asFormDataPart(
                            partName = "liveness_images",
                            mediaType = "image/jpeg",
                        )
                    },
                    partnerParams = convertNullableMapToNonNull(partnerParams),
                    callbackUrl = callbackUrl,
                    sandboxResult = sandboxResult?.toInt(),
                ).toResponse()
        },
        callback = callback,
    )

    override fun getDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        callback: (Result<FlutterDocumentVerificationJobStatusResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.getDocumentVerificationJobStatus(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun getBiometricKycJobStatus(
        request: FlutterJobStatusRequest,
        callback: (Result<FlutterBiometricKycJobStatusResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.getBiometricKycJobStatus(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun getEnhancedDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        callback: (Result<FlutterEnhancedDocumentVerificationJobStatusResponse>) -> Unit,
    ) = launch(
        work = {
            SmileID.api.getEnhancedDocumentVerificationJobStatus(request.toRequest()).toResponse()
        },
        callback = callback,
    )

    override fun getProductsConfig(
        request: FlutterProductsConfigRequest,
        callback: (Result<FlutterProductsConfigResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.getProductsConfig(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun getValidDocuments(
        request: FlutterProductsConfigRequest,
        callback: (Result<FlutterValidDocumentsResponse>) -> Unit,
    ) = launch(
        work = { SmileID.api.getValidDocuments(request.toRequest()).toResponse() },
        callback = callback,
    )

    override fun getServices(callback: (Result<FlutterServicesResponse>) -> Unit) = launch(
        work = { SmileID.api.getServices().toResponse() },
        callback = callback,
    )

    override fun pollSmartSelfieJobStatus(
        request: FlutterJobStatusRequest,
        interval: Long,
        numAttempts: Long,
        callback: (Result<FlutterSmartSelfieJobStatusResponse>) -> Unit,
    ) = launch(
        work = {
            pollJobStatus(
                apiCall = SmileID.api::pollSmartSelfieJobStatus,
                request = request.toRequest(),
                interval = interval,
                numAttempts = numAttempts,
                transform = { it.toResponse() },
            )
        },
        callback = callback,
    )

    override fun pollDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        interval: Long,
        numAttempts: Long,
        callback: (Result<FlutterDocumentVerificationJobStatusResponse>) -> Unit,
    ) = launch(
        work = {
            pollJobStatus(
                apiCall = SmileID.api::pollDocumentVerificationJobStatus,
                request = request.toRequest(),
                interval = interval,
                numAttempts = numAttempts,
                transform = { it.toResponse() },
            )
        },
        callback = callback,
    )

    override fun pollBiometricKycJobStatus(
        request: FlutterJobStatusRequest,
        interval: Long,
        numAttempts: Long,
        callback: (Result<FlutterBiometricKycJobStatusResponse>) -> Unit,
    ) = launch(
        work = {
            pollJobStatus(
                apiCall = SmileID.api::pollBiometricKycJobStatus,
                request = request.toRequest(),
                interval = interval,
                numAttempts = numAttempts,
                transform = { it.toResponse() },
            )
        },
        callback = callback,
    )

    override fun pollEnhancedDocumentVerificationJobStatus(
        request: FlutterJobStatusRequest,
        interval: Long,
        numAttempts: Long,
        callback: (Result<FlutterEnhancedDocumentVerificationJobStatusResponse>) -> Unit,
    ) = launch(
        work = {
            pollJobStatus(
                apiCall = SmileID.api::pollEnhancedDocumentVerificationJobStatus,
                request = request.toRequest(),
                interval = interval,
                numAttempts = numAttempts,
                transform = { it.toResponse() },
            )
        },
        callback = callback,
    )

    private suspend fun <RequestType, ResponseType, FlutterResponseType> pollJobStatus(
        apiCall: suspend (RequestType, Duration, Int) -> Flow<ResponseType>,
        request: RequestType,
        interval: Long,
        numAttempts: Long,
        transform: (ResponseType) -> FlutterResponseType,
    ): FlutterResponseType = try {
        val response =
            withContext(Dispatchers.IO) {
                apiCall(request, interval.milliseconds, numAttempts.toInt())
                    .map { transform(it) }
                    .single()
            }
        response
    } catch (e: Exception) {
        throw e
    }

    /**
     * https://stackoverflow.com/a/62206235
     *
     * We can get the context in a ActivityAware way, without asking users to pass the context when
     * calling "initialize" on the sdk
     */
    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        this.activity = binding.activity
    }

    override fun onDetachedFromActivityForConfigChanges() {}

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {}

    override fun onDetachedFromActivity() {}
}

/**
 * Launches a new coroutine in the specified dispatcher (IO by default) and returns the result to
 * the callback. Used for launching coroutines from Dart.
 */
private fun <T> launch(
    work: suspend () -> T,
    callback: (Result<T>) -> Unit,
    scope: CoroutineScope = CoroutineScope(Dispatchers.IO),
) {
    val handler =
        CoroutineExceptionHandler { _, throwable ->
            callback.invoke(Result.failure(throwable))
        }
    scope.launch(handler) {
        callback.invoke(Result.success(work()))
    }
}
